File: all-code/701-800/778swimInWater.py
# 在一个 n x n 的整数矩阵 grid 中，每一个方格的值 grid[i][j] 表示位置 (i, j) 的平台高度。
#
# 当开始下雨时，在时间为 t 时，水池中的水位为 t 。你可以从一个平台游向四周相邻的任意一个平台，但是前提是此时水位必须同时淹没这两个平台。假定你可以瞬间移动无限距离，也就是默认在方格内部游动是不耗时的。当然，在你游泳的时候你必须待在坐标方格里面。
#
# 你从坐标方格的左上平台 (0，0) 出发。返回 你到达坐标方格的右下平台 (n-1, n-1) 所需的最少时间 。
#
#
#
# 示例 1:
#
#
#
# 输入: grid = [[0,2],[1,3]]
# 输出: 3
# 解释:
# 时间为0时，你位于坐标方格的位置为 (0, 0)。
# 此时你不能游向任意方向，因为四个相邻方向平台的高度都大于当前时间为 0 时的水位。
# 等时间到达 3 时，你才可以游向平台 (1, 1). 因为此时的水位是 3，坐标方格中的平台没有比水位 3 更高的，所以你可以游向坐标方格中的任意位置
# 示例 2:
#
#
#
# 输入: grid = [[0,1,2,3,4],[24,23,22,21,5],[12,13,14,15,16],[11,17,18,19,20],[10,9,8,7,6]]
# 输出: 16
# 解释: 最终的路线用加粗进行了标记。
# 我们必须等到时间为 16，此时才能保证平台 (0, 0) 和 (4, 4) 是连通的
#
#
# 提示:
#
# n == grid.length
# n == grid[i].length
# 1 <= n <= 50
# 0 <= grid[i][j] < n2
# grid[i][j] 中每个值 均无重复


from typing import List
from heapq import *

class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        n = len(grid)
        hq = [[grid[0][0], 0, 0]]
        vis = [[0] * n for _ in range(n)]
        vis[0][0] = 1
        ans = grid[0][0]
        dir = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        while len(hq):
            t, i, j = heappop(hq)
            ans = max(ans, t)
            for u, v in dir:
                x, y = i + u, j + v
                if 0 <= x < n and 0 <= y < n and not vis[x][y]:
                    if x == n - 1 and y == n - 1:
                        return max(grid[x][y], ans)
                    vis[x][y] = 1
                    heappush(hq, [grid[x][y], x, y])
        return ans



so = Solution()
print(so.swimInWater(grid = [[0]]))  # 0
print(so.swimInWater(grid = [[3,2],[0,1]]))  # 3
print(so.swimInWater(grid = [[0,2],[1,3]]))  # 3
print(so.swimInWater(grid = [[0,1,2,3,4],[24,23,22,21,5],[12,13,14,15,16],[11,17,18,19,20],[10,9,8,7,6]]))  # 16

File: c++/src/1-999/778swimInWater.cpp
#include "lc_pub.h"

using namespace std;

class UnionFind {
    vector<int> fa; // 代表元
    vector<int> sz; // 集合大小

public:
    int cc; // 连通块个数

    UnionFind(int n) : fa(n), sz(n, 1), cc(n) {
        // 一开始有 n 个集合 {0}, {1}, ..., {n-1}
        // 集合 i 的代表元是自己，大小为 1
        ranges::iota(fa, 0); // iota(fa.begin(), fa.end(), 0);
    }

    // 返回 x 所在集合的代表元
    // 同时做路径压缩，也就是把 x 所在集合中的所有元素的 fa 都改成代表元
    int find(int x) {
        // 如果 fa[x] == x，则表示 x 是代表元
        if (fa[x] != x) {
            fa[x] = find(fa[x]); // fa 改成代表元
        }
        return fa[x];
    }

    // 判断 x 和 y 是否在同一个集合
    bool is_same(int x, int y) {
        // 如果 x 的代表元和 y 的代表元相同，那么 x 和 y 就在同一个集合
        // 这就是代表元的作用：用来快速判断两个元素是否在同一个集合
        return find(x) == find(y);
    }

    // 把 from 所在集合合并到 to 所在集合中
    // 返回是否合并成功
    bool merge(int from, int to) {
        int x = find(from), y = find(to);
        if (x == y) { // from 和 to 在同一个集合，不做合并
            return false;
        }
        fa[x] = y; // 合并集合。修改后就可以认为 from 和 to 在同一个集合了
        sz[y] += sz[x]; // 更新集合大小（注意集合大小保存在代表元上）
        // 无需更新 sz[x]，因为我们不用 sz[x] 而是用 sz[find(x)] 获取集合大小，但 find(x) == y，我们不会再访问 sz[x]
        cc--; // 成功合并，连通块个数减一
        return true;
    }

    // 返回 x 所在集合的大小
    int get_size(int x) {
        return sz[find(x)]; // 集合大小保存在代表元上
    }
};

class Solution {
public:
    int swimInWater(vector<vector<int>>& grid) {
        int dir[4][2] = {{0, 1},{1,0},{0,-1},{-1,0}};
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
        int r=grid.size(),c=grid[0].size();
        UnionFind uf(r * c);
        unordered_set<int> vis;
        for (int i=0;i<r;i++) {
            for (int j=0;j<c;j++) {
                pq.push({grid[i][j],i,j});
                
                // for (int k=0;k<4;k++) {
                //     auto [dx, dy] = dir[k];
                //     int x=i+dx,y=j+dy;
                //     if (0<=x&&x<r&&0<=y&&y<c&&vis.find(x*c+y)!=vis.end())
                //         uf.merge(i*c+j,x*c+y);
                // }
            }
        }
        while (true) {
            auto& arr = pq.top();
            int ans=arr[0], i=arr[1], j=arr[2];
            for (int k=0;k<4;k++) {
                auto [dx, dy] = dir[k];
                int x=i+dx,y=j+dy;
                if (0<=x&&x<r&&0<=y&&y<c&&vis.find(x*c+y)!=vis.end())
                    uf.merge(i*c+j,x*c+y);
            }
            vis.insert(i * c + j);
            pq.pop();
            
            if (uf.find(0)==uf.find(r*c-1)) return ans;
        }

    }
};

int main()
{
    std::cout<<"test let us start! %s" << __cplusplus <<std::endl;
    auto arr=parseGrid("[[0,2],[1,3]]");
    Solution so;
    cout << so.swimInWater(arr) <<endl;
    return 0;
}